api/callbacks.go

```diff
@@ -46,14 +46,27 @@ func buildDB(kv KVStore) C.DB {
 }
 
 //export cGet
-func cGet(ptr *C.db_t, key C.Buffer, val C.Buffer) i64 {
+func cGet(ptr *C.db_t, key C.Buffer, val *C.Buffer) (ret i32) {
+	// If the SDK panics, return -1 to inform the rust side something failed
```

Comment:
Can you please give me a pointer to the methods in the SDK that will tell me how much gas was used by the read? we will have to add a description of them to the `KVStore` interface if they are methods on cosmos's storage type. I think it will be useful if this function returned an i64, which represents the amount of gas consumed when positive, and represents an error when negative.

Reply:
SDK will panic if it runs out of gas.
That is desired behavior by the designers of CosmosSDK and they catch that particular panic message in `baseapp/run_tx`.

We don't want to start getting into passing this info into rust, to count SDK gas for now. That can be revisited later with resolving https://github.com/CosmWasm/go-cosmwasm/issues/29 but out of scope for this PR.

Reply:
I see... I was not aware of panics being a part of the expected behaviour here.
We should check what type is sent in the panic, and if it is an "out of gas" event, that needs to be correctly passed into Rust (we can use an opaque pointer to the object in the panic for example), return the error inside the Rust code all the way back to the Go code that called us, and then panic again on the Go side with that same object. I imagine that the SDK knows to expect that same object higher in the stack.
Should i try and implement such a thing? as it's implemented right now the rust side will panic when this happens, and that panic is later caught in `lib.rs` but then the Go code doesn't know the failure happened because of the out-of-gas event.

Reply:
Let's look at this as part of another PR. Good to get this in, then test `wasmd` behavior with this.

Reply:
Should i open an issue about this, then?

Reply:
Yes, please

Reply:
I added a detailed comment on this topic on this issue:
https://github.com/CosmWasm/go-cosmwasm/issues/29